### PR TITLE
Better container integration testing automation

### PR DESCRIPTION
**Pull request by HenryL27**
- Jupyter testing
- Create compose stack as a fixture
- provide a tag to run tests against

### CODE REVIEW

File: apps/integration/integration/automation/runtests.sh [hunk 1]
```diff
@@ -39,5 +39,14 @@ handle_outputs() {
   echo "Yep, definitely handling test outputs. That's what this function does" >&2
 }
 
+runtests() {
+  docker system prune -f --volumes
+  docker compose up reset
+  poetry run pytest apps/integration/ -p integration.conftest --noconftest --docker-tag latest_rc
+  # this is a complicated command, so: ^                        ^            ^ test against containers tagged latest_rc
+  #                                    |                     don't load conftest at pytest runtime; it's already loaded
+  #                                     load conftest with plugins, to capture the custom command line arg --docker-tag
```
Comment (alexaryn) [inline apps/integration/integration/automation/runtests.sh:48]: Nice hack.  Not sure the ASCII art translates to narrower screens.

File: apps/integration/integration/containers/stack.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+from typing import List
+
+
+def docker_compose(services: List[str] = []):
+    """
+    Get a docker compose object that controls some set of the sycamore stack
+    :param services: the list of services to control. [] -> all services
+    :return: the docker compose object
+    """
+    if len(services) > 0:
```
Comment (alexaryn) [inline apps/integration/integration/containers/stack.py:13]: I'd just have `if not services:` so people can pass in None, too.  Reads better, too.

File: apps/integration/integration/containers/stack.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+        return DockerCompose(SYCAMORE_ROOT, services=services)
+    else:
+        return DockerCompose(SYCAMORE_ROOT)
+
+
+def set_version_tag(tag: str):
+    with open(SYCAMORE_ROOT / ".env", "r") as f:
+        lines = f.readlines()
+    lines[0] = f"VERSION={tag}\n"
```
Comment (alexaryn) [inline apps/integration/integration/containers/stack.py:22]: This seems remarkably delicate.  What if somebody puts a comment at the top of .env?

File: apps/integration/integration/ingests/index.py [hunk 4]
```diff
@@ -1,11 +1,12 @@
 import pytest
 from integration.ingests.crawler import HttpCrawlerIndex
+from integration.ingests.jupyter import JupyterIndex
 
+INGEST_PROFILES = ["crawler-http-one", "crawler-http-all", "jupyter-default-prep", "jupyter-dev-example"]
```
Comment (alexaryn) [inline apps/integration/integration/ingests/index.py:5]: Do we care about sycamore_demo?

File: apps/integration/integration/ingests/index.py [hunk 5]
```diff
@@ -25,6 +26,13 @@ def ingested_index(opensearch_client, container_handles, ingest_profile):
...
             profile="sort-all", opensearch=opensearch_client, importer=container_handles["importer"]
         )
+    elif ingest_profile == "jupyter-default-prep":
+        index_ctx = JupyterIndex(
+            nb_name="default-prep-script.ipynb", opensearch=opensearch_client, jupyter=container_handles["jupyter"]
+        )
+    elif ingest_profile == "jupyter-dev-example":
+        index_ctx = JupyterIndex(
+            nb_name="jupyter_dev_example.ipynb", opensearch=opensearch_client, jupyter=container_handles["jupyter"]
+        )
```
Comment (alexaryn) [inline apps/integration/integration/ingests/index.py:36]: Could this stuff be data-driven or based on patterns?  A bunch of if-then-else logic isn't ideal when the code is so similar.

File: apps/integration/integration/ingests/jupyter.py [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+}
+
+
+class JupyterIndex:
+    def __init__(self, nb_name: str, opensearch: OpenSearch, jupyter: Container):
+        self._nb_name = nb_name
+        self._opensearch = opensearch
+        self._jupyter = jupyter
+        if nb_name not in JUPYTER_NB_INFO:
+            raise ValueError(f"Unrecognized notebook name: {nb_name}")
```
Comment (alexaryn) [inline apps/integration/integration/ingests/jupyter.py:19]: I'd expect to see parameter validation come first.

**Review by alexaryn — APPROVED** [review]
Some comments; no blockers.

File: apps/integration/integration/containers/stack.py (hunk 3 above)
Comment (HenryL27) [reply apps/integration/integration/containers/stack.py:22]: well, don't! but also yeah will fix

File: apps/integration/integration/ingests/index.py (hunk 4 above)
Comment (HenryL27) [reply apps/integration/integration/ingests/index.py:5]: If by that you mean `jupyter/sycamore_demo.ipynb`, then it isn't in the container, so no. If there's some other sycamore_demo then I don't know what it is

File: apps/integration/integration/ingests/index.py (hunk 5 above)
Comment (HenryL27) [reply apps/integration/integration/ingests/index.py:36]: Could do something like that, sure (although I think python's `match` statement is only a thing in 3.12, and you really want that for pattern matching stuff... pretty sure until 3.12 if elif is how you do that). I'm not sure we're ever going to have enough ingest profiles that it's worth doing anything more complicated than simply enumerating them. remember, each one of these is a full sycamore ingest job + 48 opensearch queries, so the cost of adding to this section is not in the complexity/duplicativeness of this section

File: apps/integration/integration/automation/runtests.sh (hunk 1 above)
Comment (HenryL27) [reply apps/integration/integration/automation/runtests.sh:48]: I'll let you take up the 80 versus 120 argument with someone else.

**Pull request merged**